### PR TITLE
INTL-1332: Add import smarts to codemod

### PR DESCRIPTION
**Pull request by chetansinghchoudhary-wk**
### Problem: 
  - Add the ability for the codemod to decide how the intl class import statement will be formatted, ie. package format vs. relative format. It should behave something like this:
      - If the file has only **package** imports, go with a **package** format
      - If the file has only **relative** imports, go with a **relative** format
      - If the file has a mix of both, go with the **package** format
      - If the file has no imports, go with the **package** format
  
 ### Solution:
  - Determine where to insert the import statement. The `_insertionLocationForPackageImport` function is responsible for this.
  - This function inspects the existing import directives in the file and calculates the appropriate insertion point for the `intl` import. 
  - Tracked whether the file has **package** or **relative** imports by iterating through the existing import directives.
  - After determining the insertion location and identifying existing imports, decide on the import format. This is done in the `decideImportFormat` function.
If there are **package** imports and no **relative** imports, it selects the **package** format.
Otherwise, it selects the **relative** format.

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:52]: I don't think we know for certain that it will be two levels up. A file could be in lib/src, lib/src/project, lib/src/project/components, lib/src/project/components/report_component/redux, etc.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:145]: A relative import doesn't necessarily start with "..", because it could be importing at the same level, for example
import 'other_file.dart';

You can ask a URI if it's relative or absolute, which I think would be better.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:141]: This needs to only check package: URLS that are for *this* package. e.g.

import 'package:path/path.dart';
import '../foo.dart';

In this case it is using relative imports only.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:140]: It's probably preferable to use a regular for-loop for this. They're easier to debug in Dart. Or you could write it as two "any" loops. 

hasPackageImports = imports.any((importDirective) => ...
hasRelativeImports = imports.any(...

File: lib/src/intl_suggestors/intl_importer.dart
Comment (chetansinghchoudhary-wk) [reply lib/src/intl_suggestors/intl_importer.dart:52]: Yes, files may be in different packages.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (chetansinghchoudhary-wk) [reply lib/src/intl_suggestors/intl_importer.dart:49]: update with `bool` type

File: lib/src/intl_suggestors/intl_importer.dart
Comment (chetansinghchoudhary-wk) [reply lib/src/intl_suggestors/intl_importer.dart:140]: Updated with `for` loop

**Review by akshaybhardwaj-wk — APPROVED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:47]: Does this (computing a relative path between a URI and a file system path) work in general? I think you could do this more simply as
`intlDirectory = path.join(context.root, 'lib/src/intl/${projectName}_intl.dart');`
`path.relative(intlDirectory, path.current)`
and then you don't need to manually compute relativeImportPath.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:54]: How does this fix the ordering problem? It seems like this is always finding the insert position as if the import path was relative. Shouldn't you compute the insertInfo after you have the importStatement?

File: test/intl_suggestors/intl_importer_test.dart
Comment (alanknight-wk) [inline test/intl_suggestors/intl_importer_test.dart:150]: The point of the rather complicated _insertLocationForPackageImport is that it tries to put it in alphabetical order. This seems like it's just making it so that for the case of package: imports the new import will always comes first (because it computes the position as if it started with '..') and then changes the test cases so that coming first is expected. I think these should go back to the way they were, and the computation should work on the actual form of the URI that we're going to use.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:174]: This probably should have a name that indicates what the boolean result means. Like shouldUsePackageImports (but then maybe the parameter by that name needs to change). And a comment to describe the decision.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [reply lib/src/intl_suggestors/intl_importer.dart:154]: This is still checking ALL package imports, not just the ones in this package. And the decision process is that we use package imports if there are ANY package imports. So we will not use relative imports if there are imports from any other packages, which is wrong.

File: test/intl_suggestors/intl_importer_test.dart
Comment (alanknight-wk) [inline test/intl_suggestors/intl_importer_test.dart:383]: This doesn't seem like the correct test. If we are using relative imports within the package (the only place we can) then we should be using a relative import for the _intl.dart file. Otherwise we might as well not bother doing this, because there are very few files that will only have imports from within their own package.

**Review by alanknight-wk — CHANGES_REQUESTED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (chetansinghchoudhary-wk) [reply lib/src/intl_suggestors/intl_importer.dart]: Updated as you suggested.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (chetansinghchoudhary-wk) [reply lib/src/intl_suggestors/intl_importer.dart]: Updated as you suggested.

File: test/intl_suggestors/intl_importer_test.dart
Comment (chetansinghchoudhary-wk) [reply test/intl_suggestors/intl_importer_test.dart:150]: Updated the test cases and

File: test/intl_suggestors/intl_importer_test.dart
Comment (chetansinghchoudhary-wk) [reply test/intl_suggestors/intl_importer_test.dart:383]: The import format is "package" when package and relative imports exist.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart]: > Does this (computing a relative path between a URI and a file system path) work in general? I think you could do this more simply as `intlDirectory = path.join(context.root, 'lib/src/intl/${projectName}_intl.dart');` `path.relative(intlDirectory, path.current)` and then you don't need to manually compute relativeImportPath.

I have a one-line `final currentDirectory = Directory.current`; before   `final intlDirectory = path.join(currentDirectory.path, 'lib/src/intl/${projectName}_intl.dart');`

File: test/intl_suggestors/intl_importer_test.dart
Comment (alanknight-wk) [inline test/intl_suggestors/intl_importer_test.dart:64]: Probably delete the extra blank lines that were introduced.

File: test/intl_suggestors/intl_importer_test.dart
Comment (alanknight-wk) [inline test/intl_suggestors/intl_importer_test.dart:157]: There should be a test case for a package import in the same package. All these tests produce relative imports.

**Review by alanknight-wk — CHANGES_REQUESTED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart]: Got it I have updated the and direct use this `Directory.current.path` instead of taking in variable 
  `final intlDirectory = path.join(Directory.current.path, intlFilePath);`

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart]: code updated

File: test/intl_suggestors/intl_importer_test.dart
Comment (akshaybhardwaj-wk) [reply test/intl_suggestors/intl_importer_test.dart]: updated the code

**Review by dustinlessard-wf — DISMISSED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart:45]: Code updated

**Review by dustinlessard-wf — APPROVED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:147]: This could be more concisely written as something like

bool hasRelativeImports(imports) imports.any((importDirective) => importDirective.uri != null && importDirective.uri.scheme != 'package' && importDirective.uri.scheme != 'dart');

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:150]: I think it's also possible for the scheme to be 'dart:'. I handled that in the comment above.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:116]: This name is not clear, as below you're now setting it to false if there are any relative imports at all. Should maybe be hasRelativeImports?

File: test/intl_suggestors/intl_importer_test.dart
Comment (alanknight-wk) [inline test/intl_suggestors/intl_importer_test.dart:69]: added blank line

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:46]: These should be using context.root, not Directory.current

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart:116]: This `bool` is for enabling true condition where we want imports with the package and if it is not `package` then we are setting it to false.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart:150]: The above code has handled all the cases, for specific `uri.scheme != 'package'` I have done it separately and `hasPackageImports=false `so It working here and above condition as well.  also the `bool usePackageImports` are getting updated in return.

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart:147]: Code updated like 
```
 for (final importDirective in imports) {
    final uriContent = importDirective.uri.stringValue;
    if (uriContent != null) {
      final uri = Uri.parse(uriContent);
      hasPackageImports = !imports.any((importDirective) =>
          uri != null && uri.scheme != 'package' && uri.scheme != 'dart:');
    }
  }
```

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [reply lib/src/intl_suggestors/intl_importer.dart:116]: OK. Maybe it should be hasOnlyPackageImports? Or hasAllPackageImports?

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [reply lib/src/intl_suggestors/intl_importer.dart:147]: Now you're looping over imports twice. If you have the added line for hasPackageImports =, then you don't need the outer for loop.

**Review by alanknight-wk — CHANGES_REQUESTED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart]: code updated

File: lib/src/intl_suggestors/intl_importer.dart
Comment (alanknight-wk) [inline lib/src/intl_suggestors/intl_importer.dart:147]: Very close. But these lines are not necessary This is taking something that returns a Uri, turning it into a string, then parsing it so you can ask it for its scheme. Just get the uri and check if it's null and check the scheme.

**Review by alanknight-wk — CHANGES_REQUESTED** [review]

File: lib/src/intl_suggestors/intl_importer.dart
Comment (akshaybhardwaj-wk) [reply lib/src/intl_suggestors/intl_importer.dart:147]: Initially, we are writing code like [this](https://github.com/Workiva/over_react_codemod/pull/241/commits/eaa80fded45b8034145563c9975131751c3ba8b8#diff-228c7d433e4325d61c63d9f892a715c3dadd37820e9f5a0e2c38ca3677c2ac05R150) 
` final uriContent = importDirective.uriContent;`
but after merge conflicts resolve there is updating in Analyzer 5 `uriContent` is archived and we are using `importDirective.uri.stringValue` 
and if we check scheme with uri like this 

```
hasOnlyPackageImports = !imports.any((importDirective) {
    final uri = importDirective.uri;
    return uri != null && (uri.scheme != 'package' && uri.scheme != 'dart:');
  });

```
then it's giving error `error: The getter 'scheme' isn't defined for the type 'StringLiteral'. (undefined_getter at [over_react_codemod] lib/src/intl_suggestors/intl_importer.dart:158)`

Comment (alanknight-wk) [reply lib/src/intl_suggestors/intl_importer.dart:147]: Ah, ok. My mistake. I assumed importDirective.uri was returning a Uri, since we needed to convert it to a string. It's a StringLiteral, so this code is good.

**Review by alanknight-wk — APPROVED** [review]

**Comment by alanknight-wk** [comment]
QA
The code seems reasonable now, but I ran it for QA on a few repos and it didn't do what I expected.
graph_app - no changes, but they have the codemod in their CI now, so I didn't expect any, and nothing broke.
doc_plat_client - only one change, but it was to a file with no other imports from that package, and it made it a relative import, which doesn't seem right as the default
wdesk_sdk - The oddest one. In lib/src/analytics/shell_analytics.dart there was only one previous import, to another package, it added the intl import as package: But in lib/src/experiences/configs/wdata_experience_config.dart, which did have package: imports in the current package it added it as relative.

**Comment by akshaybhardwaj-wk** [comment]
> lib/src/experiences/configs/wdata_experience_config.dart

check with wdesk_sdk 
1. lib/src/analytics/shell_analytics.dart In this file I am getting package import 
2. lib/src/experiences/configs/wdata_experience_config.dart In this I am getting relative because this file has both `dart:` and `package`  import we have to handle this case and also create a needed test case .

**Comment by akshaybhardwaj-wk** [comment]
> > lib/src/experiences/configs/wdata_experience_config.dart
> 
> check with wdesk_sdk
> 
> 1. lib/src/analytics/shell_analytics.dart In this file I am getting package import
> 2. lib/src/experiences/configs/wdata_experience_config.dart In this I am getting relative because this file has both `dart:` and `package`  import we have to handle this case and also create a needed test case .

handle this scenario and update test case but it's getting failed

**Review by alanknight-wk — APPROVED** [review]

**Comment by alanknight-wk** [comment]
QA +1 CI passes

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**
